src/clients/base.rs

```diff
@@ -0,0 +1,1054 @@
+    /// - include_external: Optional.Possible values: audio. If
+    ///   include_external=audio is specified the response will include any
+    ///   relevant audio content that is hosted externally.  
+    ///
+    /// [Reference](https://developer.spotify.com/documentation/web-api/reference/#category-search)
+    async fn search(
+        &self,
+        q: &str,
+        _type: &SearchType,
+        market: Option<&Market>,
+        include_external: Option<&IncludeExternal>,
+        limit: Option<u32>,
+        offset: Option<u32>,
+    ) -> ClientResult<SearchResult> {
+        let limit = limit.map(|s| s.to_string());
+        let offset = offset.map(|s| s.to_string());
+        let params = build_map! {
+            "q": q,
+            "type": _type.as_ref(),
+            optional "market": market.map(|x| x.as_ref()),
+            optional "include_external": include_external.map(|x| x.as_ref()),
+            optional "limit": limit.as_deref(),
+            optional "offset": offset.as_deref(),
+        };
+
+        let result = self.endpoint_get("search", &params).await?;
+        convert_result(&result)
+    }
+
+    /// Get Spotify catalog information about an album's tracks.
+    ///
+    /// Parameters:
+    /// - album_id - the album ID, URI or URL
+    /// - limit  - the number of items to return
+    /// - offset - the index of the first item to return
+    ///
+    /// See [`Self::album_track_manual`] for a manually paginated version of
+    /// this.
+    ///
+    /// [Reference](https://developer.spotify.com/documentation/web-api/reference/#endpoint-get-an-albums-tracks)
+    fn album_track<'a>(
+        &'a self,
+        album_id: &'a AlbumId,
+    ) -> Paginator<'_, ClientResult<SimplifiedTrack>> {
+        paginate(
+            move |limit, offset| self.album_track_manual(album_id, Some(limit), Some(offset)),
+            self.get_config().pagination_chunks,
+        )
+    }
+
+    /// The manually paginated version of [`Self::album_track`].
+    async fn album_track_manual(
+        &self,
+        album_id: &AlbumId,
+        limit: Option<u32>,
+        offset: Option<u32>,
+    ) -> ClientResult<Page<SimplifiedTrack>> {
+        let limit = limit.map(|s| s.to_string());
+        let offset = offset.map(|s| s.to_string());
+        let params = build_map! {
+            optional "limit": limit.as_deref(),
+            optional "offset": offset.as_deref(),
+        };
+
+        let url = format!("albums/{}/tracks", album_id.id());
+        let result = self.endpoint_get(&url, &params).await?;
+        convert_result(&result)
+    }
+
+    /// Gets basic profile information about a Spotify User.
+    ///
+    /// Parameters:
+    /// - user - the id of the usr
+    ///
+    /// [Reference](https://developer.spotify.com/documentation/web-api/reference/#endpoint-get-users-profile)
+    async fn user(&self, user_id: &UserId) -> ClientResult<PublicUser> {
```

Comment:
This endpoint should be put into `oauth.rs` instead of `base.rs`, since it needs a valid access token obtained by authorization.